icu4c/source/i18n/ucal.cpp

```diff
@@ -628,6 +628,27 @@ ucal_getCanonicalTimeZoneID(const char16_t* id, int32_t len,
     return reslen;
 }
 
+U_DRAFT int32_t U_EXPORT2
+ucal_getIanaTimeZoneID(const char16_t* id, int32_t len,
+                        char16_t* result, int32_t resultCapacity, UErrorCode* status)
+{
+    if (status == 0 || U_FAILURE(*status)) {
+        return 0;
+    }
+    if (id == 0 || len == 0 || result == 0 || resultCapacity <= 0) {
+        *status = U_ILLEGAL_ARGUMENT_ERROR;
+        return 0;
+    }
+    int32_t reslen = 0;
```

Comment:
I thnik we only need
```
    UnicodeString ianaID;
    return TimeZone::getIanaID(UnicodeString(id, len), ianaID, *status)
                  .extract(result, resultCapacity, *status);
```
Since both getIanaID and extract should check *status and return immediately anyway.

Reply:
Will check it again.

Reply:
And if id == 0 || len == 0  then UnicodeString(id, len) will produce an empty string which will presumably cause TimeZone::getIanaID to fail with an error.

So yes, Frank's suggestion  probably simplifies things. But... the current implementation still works fine. Could be simplified in a future PR.